siliconcompiler/tools/yosys/syn_asic.tcl

```diff
@@ -52,27 +44,59 @@ foreach item $sc_scenarios {
 # read_liberty calls for it to not affect synthesis results.
 yosys hierarchy -top $sc_design
 
-yosys synth "-flatten" -top $sc_design
+# Mark modules to keep from getting removed in flattening
+if {[dict exists $sc_cfg tool $sc_tool var $sc_step $sc_index preserve_modules]} {
+    foreach module [dict get $sc_cfg tool $sc_tool var $sc_step $sc_index preserve_modules] {
+        yosys select -module $module
+        yosys setattr -mod -set keep_hierarchy 1
+        yosys select -clear
+    }
+}
+
+set synth_args ""
+if {[dict get $sc_cfg tool $sc_tool var $sc_step $sc_index flatten] == "True"} {
+    append synth_args "-flatten "
+}
+yosys synth {*}$synth_args -top $sc_design
 
 yosys opt -purge
 
 ########################################################
 # Technology Mapping
 ########################################################
-if [dict exists dict get $sc_cfg library $sc_mainlib asic "file" $sc_tool techmap] {
-    set sc_techmap     [dict get $sc_cfg library $sc_mainlib asic "file" $sc_tool techmap]
-    foreach mapfile $sc_techmap {
-	yosys techmap -map $mapfile
+proc post_techmap { { opt_args "" } } {
+    # perform techmap in case previous techmaps introduced constructs that need techmapping
+    yosys techmap
+    # Quick optimization
+    yosys opt {*}$opt_args -purge
+}
+if {[dict get $sc_cfg tool $sc_tool var $sc_step $sc_index map_adders] != "False"} {
+    set sc_adder_techmap [lindex [dict get $sc_cfg library $sc_mainlib asic "file" $sc_tool addermap] 0]
+    # extract the full adders
+    yosys extract_fa
+    # map full adders
+    yosys techmap -map $sc_adder_techmap
+    post_techmap -fast
+}
+
+if [dict exists $sc_cfg tool $sc_tool var $sc_step $sc_index techmap] {
+    foreach mapfile [dict get $sc_cfg tool $sc_tool var $sc_step $sc_index techmap] {
+        yosys techmap -map $mapfile
+        post_techmap -fast
     }
 }
 
-#TODO: Fix better
-set libcorner    [lindex [dict get $sc_cfg constraint [lindex $sc_scenarios 0] libcorner] 0]
-set mainlib      [dict get $sc_cfg library $sc_mainlib model timing $sc_delaymodel $libcorner]
+if {[dict get $sc_cfg tool $sc_tool var $sc_step $sc_index autoname] == "True"} {
+    # use autoname to preserve some design naming
+    # by doing it before dfflibmap the names will be slightly shorter since they will
+    # only contain the $DFF_P names vs. the full library name of the associated flip-flop
+    yosys autoname
+}
 
-yosys dfflibmap -liberty $mainlib
+yosys dfflibmap -liberty "inputs/dff_liberty.lib"
```

Comment:
This hard coded name seems problematic.

Reply:
What would you suggest? I could pass it in via the schema and set the name in python, would that be better?

Reply:
changed it, such that the path is set by the python script, however, it is still a constant (just now it's controlled in one place and from the python tool driver)